src/api/ruleMapping/ruleMapping.service.ts

```diff
@@ -0,0 +1,160 @@
+import { Injectable } from '@nestjs/common';
+import { Node, Edge, TraceObject } from './ruleMapping.interface';
+import { DocumentsService } from '../documents/documents.service';
+
+@Injectable()
+export class RuleMappingService {
+  // Extract all inputs from a list of nodes
+  extractInputs(nodes: Node[]): { inputs: any[] } {
+    const inputs = nodes.flatMap((node: any) => {
+      if (node.type === 'expressionNode' && node.content?.expressions) {
+        return node.content.expressions.map((expr: { key: any; value: any }) => ({
+          key: expr.key,
+          property: expr.value,
+        }));
+      } else {
+        return (node.content?.inputs || []).map((inputField: { id: any; name: any; type: any; field: any }) => ({
+          id: inputField.id,
+          name: inputField.name,
+          type: inputField.type,
+          property: inputField.field,
+        }));
+      }
+    });
+
+    return { inputs };
+  }
+
+  // Extract all outputs from a list of nodes
+  extractOutputs(nodes: Node[]): { outputs: any[] } {
+    const outputs = nodes.flatMap((node: any) => {
+      if (node.type === 'expressionNode' && node.content?.expressions) {
+        return node.content.expressions.map((expr: { value: any; key: any }) => ({
+          key: expr.value,
+          property: expr.key,
+        }));
+      } else {
+        return (node.content?.outputs || []).map((outputField: { id: any; name: any; type: any; field: any }) => ({
+          id: outputField.id,
+          name: outputField.name,
+          type: outputField.type,
+          property: outputField.field,
+        }));
+      }
+    });
```

Comment:
Are extractInputs and extractOutputs essentially the same? Maybe unify them under the same function or at least have a function they both call so code isn't repeated. Also, make use of your types from your interface (InputField/OutputField)